src/control/lib/telemetry/telemetry.go

```diff
@@ -0,0 +1,396 @@
+		return err.Error()
+	}
+
+	return strings.TrimSpace(string(buf[:bytes.Index(buf, []byte{0})]))
+}
+
+func (sm *statsMetric) Mean() float64 {
+	return float64(sm.stats.mean)
+}
+
+func (sm *statsMetric) StdDev() float64 {
+	return float64(sm.stats.std_dev)
+}
+
+func (sm *statsMetric) SampleSize() uint64 {
+	return uint64(sm.stats.sample_size)
+}
+
+func getStatsMinInt(stats *C.struct_d_tm_stats_t) uint64 {
+	return uint64(C.get_min_int(stats))
+}
+
+func getStatsMaxInt(stats *C.struct_d_tm_stats_t) uint64 {
+	return uint64(C.get_max_int(stats))
+}
+
+func getStatsMinFloat(stats *C.struct_d_tm_stats_t) float64 {
+	return float64(C.get_min_float(stats))
+}
+
+func getStatsMaxFloat(stats *C.struct_d_tm_stats_t) float64 {
+	return float64(C.get_max_float(stats))
+}
+
+func getStatsSumInt(stats *C.struct_d_tm_stats_t) uint64 {
+	return uint64(C.get_sum_int(stats))
+}
+
+func getStatsSumFloat(stats *C.struct_d_tm_stats_t) float64 {
+	return float64(C.get_sum_float(stats))
+}
+
+func Init(parent context.Context, idx uint32) (context.Context, error) {
+	shmemRoot := C.d_tm_get_shared_memory(C.int(idx))
+	if shmemRoot == nil {
+		return nil, errors.Errorf("no shared memory segment found for idx: %d", idx)
+	}
+
+	root := C.d_tm_get_root(shmemRoot)
+	if root == nil {
+		return nil, errors.Errorf("no root node found in shared memory segment for idx: %d", idx)
+	}
+
+	handle := &handle{
+		idx:   idx,
+		shmem: shmemRoot,
+		root:  root,
+	}
+
+	return context.WithValue(parent, handleKey, handle), nil
+}
+
+func visit(hdl *handle, node *C.struct_d_tm_node_t, pathComps []string, out chan<- Metric) {
+	var next *C.struct_d_tm_node_t
+
+	if node == nil {
+		return
+	}
+	path := strings.Join(pathComps, "/")
+	name := C.GoString((*C.char)(C.d_tm_conv_ptr(hdl.shmem, unsafe.Pointer(node.dtn_name))))
+	//fmt.Printf("visiting: %q/%q (%p)\n", path, name, node)
+
+	switch node.dtn_type {
+	case C.D_TM_DIRECTORY:
+		/*if name != "" {
+			pathComps = append(pathComps, name)
+		}*/
+		next = (*C.struct_d_tm_node_t)(C.d_tm_conv_ptr(hdl.shmem, unsafe.Pointer(node.dtn_child)))
+		if next != nil {
+			visit(hdl, next, append(pathComps, name), out)
+		}
+	case C.D_TM_GAUGE:
+		out <- newGauge(hdl, path, &name, node)
+	case C.D_TM_COUNTER:
+		out <- newCounter(hdl, path, &name, node)
+	default:
+		//fmt.Printf("%s: metric type 0x%x not supported yet\n", name, node.dtn_type)
+	}
+
+	next = (*C.struct_d_tm_node_t)(C.d_tm_conv_ptr(hdl.shmem, unsafe.Pointer(node.dtn_sibling)))
+	if next != nil && next != node {
+		visit(hdl, next, pathComps, out)
+	}
+}
+
+func CollectMetrics(ctx context.Context, dirname string, out chan<- Metric) error {
+	hdl, err := getHandle(ctx)
+	if err != nil {
+		return err
+	}
+
+	node := hdl.root
+
+	/*if dirname != "/" && dirname != "" {
+		node = FindMetric(shmemRoot, dirname)
+	}*/
+
+	if node == nil {
+		return errors.Errorf("Directory or metric:[%s] was not found", dirname)
+	}
+
+	var nl *C.struct_d_tm_nodeList_t
+
+	filter := C.D_TM_ALL_NODES
```

Comment:
Just make the definition of D_TM_ALL_NODES consistent with the types of metrics you actually wanted to handle so you avoid reading and digesting metric data for items you didn't really want.

Reply:
I'd like for that constant to really mean all nodes rather than what my particular consumer is interested in. I'd rather err on the side of getting too much vs. missing some and not knowing.